<a href="https://colab.research.google.com/github/JJ-HMFIC/ML_Basic/blob/main/ML_Basic_Ch11_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

11-2 Deep CNN

In [ ]:
from tensorflow.keras import layers, models, datasets

# Set random seed for reproducibility
tf.random.set_seed(777)

# Load MNIST dataset
mnist = datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

# Normalize pixel values to be between 0 and 1
train_images, test_images = train_images / 255.0, test_images / 255.0

# Add a channel dimension to the images
train_images = train_images[..., tf.newaxis]
test_images = test_images[..., tf.newaxis]

# One-hot encode the labels
train_labels = tf.keras.utils.to_categorical(train_labels, 10)
test_labels = tf.keras.utils.to_categorical(test_labels, 10)

# Hyperparameters
learning_rate = 0.001
training_epochs = 15
batch_size = 100

# Create the model
model = models.Sequential()

# L1
model.add(layers.Conv2D(filters=32, kernel_size=(3, 3), input_shape=(28, 28, 1), activation='relu'))
model.add(layers.MaxPooling2D(pool_size=(2, 2)))

# L2
model.add(layers.Conv2D(filters=64, kernel_size=(3, 3), activation='relu'))
model.add(layers.MaxPooling2D(pool_size=(2, 2)))

# L3 fully connected
model.add(layers.Flatten())
model.add(layers.Dense(units=625, kernel_initializer=tf.keras.initializers.GlorotNormal(), activation='relu'))
model.add(layers.Dropout(0.5)) # dropout 적용 ★

# L4 Final FC
model.add(layers.Dense(units=10, kernel_initializer=tf.keras.initializers.GlorotNormal(), activation='softmax'))

# Compile the model
model.compile(optimizer=tf.optimizers.Adam(learning_rate),
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
print('Learning started. It takes some time.')
model.fit(train_images, train_labels, batch_size=batch_size, epochs=training_epochs, verbose=1)

print('Learning Finished!')

# Test the model and check accuracy
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('Test accuracy:', test_acc)

# Get one and predict
import numpy as np
r = np.random.randint(0, len(test_images))
sample_image = test_images[r].reshape(1, 28, 28, 1)
prediction = model.predict(sample_image)
print("Label:", np.argmax(test_labels[r]))
print("Prediction:", np.argmax(prediction))


Learning started. It takes some time.
Epoch 1/15
600/600 [==============================] - 35s 56ms/step - loss: 0.1770 - accuracy: 0.9460
Epoch 2/15
600/600 [==============================] - 30s 51ms/step - loss: 0.0549 - accuracy: 0.9832
Epoch 3/15
600/600 [==============================] - 31s 52ms/step - loss: 0.0412 - accuracy: 0.9876
Epoch 4/15
600/600 [==============================] - 31s 52ms/step - loss: 0.0316 - accuracy: 0.9900
Epoch 5/15
600/600 [==============================] - 38s 64ms/step - loss: 0.0263 - accuracy: 0.9916
Epoch 6/15
600/600 [==============================] - 31s 52ms/step - loss: 0.0216 - accuracy: 0.9933
Epoch 7/15
600/600 [==============================] - 30s 51ms/step - loss: 0.0178 - accuracy: 0.9942
Epoch 8/15
600/600 [==============================] - 31s 52ms/step - loss: 0.0166 - accuracy: 0.9946
Epoch 9/15
600/600 [==============================] - 30s 51ms/step - loss: 0.0145 - accuracy: 0.9952
Epoch 10/15
600/600 [=======================

Deep CNN이 기존 CNN 보다 좋은 점 : 정확도가 상승함

98.9% => 99.3%